controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,260 @@
+					oldObj := e.ObjectOld.(*v1beta1.ConstraintPodStatus)
+					newObj := e.ObjectNew.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(newObj.Status.Operations, "audit") &&
+						oldObj.Status.ObservedGeneration != newObj.Status.ObservedGeneration
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+			},
+			)).
+		Complete(r)
+}
+
+// User set gatekeeper.spec.audit.auditFromCache to Automatic, this reconcile function
+// collect requested constraints and add found kinds which is used in contraint to config.spec.sync.syncOnly
+func (r *ConstraintStatusReconciler) Reconcile(ctx context.Context,
+	request reconcile.Request,
+) (reconcile.Result, error) {
+	log := r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	log.Info("Reconciling ConstraintPodStatus and Config")
+
+	// Get config or create if not exist
+	config := &v1alpha1.Config{}
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: r.Namespace,
+		Name:      "config",
+	}, config)
+
+	if apierrors.IsNotFound(err) {
+		config = &v1alpha1.Config{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "config",
+				Namespace: r.Namespace,
+			},
+		}
+		err = r.Create(ctx, config)
+
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	contraintPodStatus := &v1beta1.ConstraintPodStatus{}
+	err = r.Get(ctx, request.NamespacedName, contraintPodStatus)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Error(err, "Cannot find any contraintStatus")
+
+			return reconcile.Result{}, nil
+		}
+		// Requeue
+		return reconcile.Result{}, err
+	}
+
+	labels := contraintPodStatus.GetLabels()
+
+	constraintKind := labels["internal.gatekeeper.sh/constraint-kind"]
+	constraintName := labels["internal.gatekeeper.sh/constraint-name"]
+
+	table := map[v1alpha1.SyncOnlyEntry]bool{}
+	// Add to table for uniqueue filtering
+	for _, entry := range config.Spec.Sync.SyncOnly {
+		table[entry] = true
+	}
+
+	constraintGVR := schema.GroupVersionResource{
+		Group:    "constraints.gatekeeper.sh",
+		Version:  "v1beta1",
+		Resource: strings.ToLower(constraintKind),
+	}
+
+	constraint, err := r.DynamicClient.Resource(constraintGVR).Get(ctx, constraintName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	constraintMatchKinds, _, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	contraintSyncOnlyEntries, err := r.getSyncOnlys(constraintMatchKinds)
+	if err != nil {
+		log.Error(err, "Error to get matching kind and apigroup")
+
+		return reconcile.Result{}, nil
+	}
+
+	for _, ce := range *contraintSyncOnlyEntries {
+		table[ce] = true
+	}
+
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+	for key := range table {
+		syncOnlys = append(syncOnlys, key)
+	}
+
+	config.Spec.Sync.SyncOnly = syncOnlys
```

Comment:
Your algorithm only ever adds syncOnly entries and does not account for constraints being deleted or being updated with a different match.

My suggestion is to have a `sync.Map` as a struct member field with the keys representing the constraints and the values being the constraint's SyncOnly values. Then after calculating the *new* SyncOnly values (i.e. add, update, or delete) during the reconcile, you can replace the constraint's SyncOnly value from the `sync.Map` and calculate what the SyncOnly entry should be.

Note that for this to work, on either the first reconcile after startup or before the controller starts up, the `sync.Map` needs to be populated.

You also may need to think about concurrency since multiple reconcilers could be updating the `sync.Map` and the Gatekeeper `Config` object. I think without any additional locking, it would just work but you may get reconciles that fail to update because the Gatekeeper `Config` object changed by another reconcile running concurrently. In this case it'd be retried and I think it'd just work.